node/core/backing/src/lib.rs

```diff
@@ -193,6 +198,23 @@ impl From<FromJob> for AllMessages {
 	}
 }
 
+impl TryFrom<AllMessages> for FromJob {
```

Comment:
The [`trait JobTrait` definition](https://github.com/paritytech/polkadot/pull/1376/files#diff-8a0bb46e1e7d7d34cfb3e47e0ce0605aR321) specifies that a job has access to a `mpsc::Sender<FromJob>`. This is to prevent jobs from sending messages they aren't supposed to.

However, we provide [helpers for certain runtime requests](https://github.com/paritytech/polkadot/pull/1376/files#diff-8a0bb46e1e7d7d34cfb3e47e0ce0605aR73-R77); those are generic over `SenderMessage`, to abstract over the fact that each impl's `ToJob` type will be distinct. These helpers need to start with an `AllMessage` instance with the actual request, and convert it into the proper `SenderMessage`. That requires either `From<AllMessages>` or `TryFrom<AllMessages>` for `SenderMessage`; I chose the latter as being more flexible.

However, I think you're right; this could use some re-architecture to be more sensible.

Reply:
One thing which will make things clearer right away is to rename `SenderMessage` to `FromJob`; they're effectively always the same type.

802981a9

Reply:
Spent a fair amount of time trying to make this work right; the results are in a5639e36017a72656b478caddcaa30e2d4e6112a.

Reverted that commit because it was adding complexity without adding benefit. I think these free functions with unintuitive trait bounds are still more generally useful than anything fancier that I've yet come up with.